### ISSUE TITLE
Should Async.async take a function with F[Unit] as the return type?

### ISSUE THREAD

**Issue by alexandru** [typelevel/cats-effect#123]
At https://github.com/typelevel/cats-effect/pull/121#issuecomment-364894096 we have a proposal to have an `Async.async` that is not side-effecting in its callback, thus with a callback that uses `F[Unit]`.

Currently we have this:

```scala
trait Async[F[_]] extends Sync[F] with LiftIO[F] {

  def async[A](k: (Either[Throwable, A] => Unit) => Unit): F[A]
}
```

Proposal is to change it to (1):

```scala
trait Async[F[_]] extends Sync[F] with LiftIO[F] {

  def async[A](k: (Either[Throwable, A] => F[Unit]) => Unit): F[A]
}
```

Or this (2):

```scala
trait Async[F[_]] extends Sync[F] with LiftIO[F] {

  def async[A](k: (Either[Throwable, A] => F[Unit]) => F[Unit]): F[A]
}
```

Arguments:

1. purity
2. a function such as [Promise.getOrWait](https://github.com/functional-streams-for-scala/fs2/blob/series/0.10/core/shared/src/main/scala/fs2/async/Promise.scala#L110) in FS2 would no longer need `Effect[F]`

**Comment by alexandru** [comment]
I'm 👎  on making this change — being not only a matter of opinion but because it's not possible.

`Async.async` is meant for wrapping side-effecting code, just like `Sync.delay`. For example the implementation of `IO.shift` is this:

```scala
  def shift(implicit F: Async[F], ec: ExecutionContext): F[Unit] = {
    F.async { (cb: Either[Throwable, Unit] => Unit) =>
      // Side-effectful API
      ec.execute(new Runnable {
        // Side-effectful signaling — after an asynchronous boundary happened!
        def run() = cb(Right(()))
      })
    }
  }
```

Describing it with any of the two alternative variants above is impossible without requiring `Effect[F]`:

```scala
  def shift(implicit F: Effect[F], ec: ExecutionContext): F[Unit] = {
    F.async { (cb: Either[Throwable, Unit] => Unit) =>
      F.delay {
        // Side-effectful API
        ec.execute(new Runnable {
          // Side-effectful signaling — after an asynchronous boundary happened!
          def run() = 
            // N.B. I have no idea if this call is correct, because I can never remember that API ;-)
            F.runAsync(cb(Right(()))).unsafeRunSync
        })
      }
    }
  }
```

This change would make the `Async` type class extremely cumbersome for its primary use-case, much less efficient and also exactly as powerful as `Effect` 😉

**Comment by SystemFw** [comment]
I think requiring `Effect` is a strong enough argument in favour of keeping it as is.

**Comment by pchlupacek** [comment]
Any strong reason why `Async` is independent from `Effect`? i.e. do you have example where `Async` could be defined for `F` but  `Effect` could not ? I cannot recall in fs2 where we have used `Async`. Could you @SystemFw ?

**Comment by alexandru** [comment]
@pchlupacek in Monix building an `Effect[Task]` requires a `Scheduler` (aka `ExecutionContext`) to be available, whereas the `Async[Task]` instance does not.

This is because the underlying execution context / scheduler in Monix gets specified at the edge, when you trigger `runAsync` and nowhere else, making `Task` truly pure. For example the `delayExecution` operation on `Task`, or the plain `Task.shift` do not require an implicit or explicit `ExecutionContext` or `Scheduler`. An overload for `Task.shift` is available, accepting a `Scheduler`, but that's there for convenience, not necessity, like is the case for `IO.shift`.

This is basically Monix's `Task.shift`:

```scala
// Look ma, no implicit ExecutionContext ;-)
val shift: Task[Unit] =
  Task.unsafeCreate { (ctx, cb) =>
    ctx.scheduler.execute(() => cb(Right(())))
  }
```

You can think of it like this: in order to trigger a `runAsync` on a Monix `Task`, you need a `Scheduler` which is part of the environment. In Haskell this would be its RTS.

---

`Async` can be implemented for streaming types, whereas `Effect` cannot be. E.g. here's the `monix.tail.Iterant` instance:

```scala
class IterantAsync[F](implicit F: Async[F]) extends Async[({type λ[α] = Iterant[F, α]})#λ] {

  def async[A](k: (Either[Throwable, A] => Unit) => Unit): Iterant[F, A] =
    Iterant.liftF(F.async(k))
}
```

We cannot implement `Effect` due to the contract, `Effect` requiring that the data type needs to stream exactly one event. N.B. whether this is a good idea or not, it's totally irrelevant, the point here being that `Async` is in fact more relaxed than `Effect` 😉

Also for `monix.tail.Iterant` I'm currently working on this operation that does not need an `Effect`:

```scala
def intervalAtFixedRate[F[_]](period: FiniteDuration)(implicit timer: Timer[F], F: Async[F]): Iterant[F, Long]
```

Currently this is specialized for `Task`, however describing a `Timer[Task]` (which would be a pure `Scheduler`) does not require an implicit `Scheduler`, due to `Task`'s nature, but it would require one for a `Timer[IO]`.

So for usage with `Task` this operation has a totally pure signature, whereas `IO` would require some sort of impure `Scheduler` in scope and make no mistake about it, this is known limitation of `IO`.

---

You can in fact describe an asynchronous file channel, based on Java NIO, just with `Async`:

```scala
abstract class AsyncFileChannel[F] {
  protected def read(dst: ByteBuffer, position: Long, cb: Either[Throwable, Unit] => Unit): Unit

  def read(dst: ByteBuffer, position: Long): F[Int] =
    F.async(cb => read(dst, position, cb))
}
```

And based on this you can implement a stream such as `monix.tail.Iterant` that reads a file:

```scala
def readFile[F](file: File)(implicit F: Async[F]): Iterant[F, Array[Byte]]
```

--- 

In fact, the interesting thing about `monix.tail.Iterant` is that it doesn't need `Effect` anywhere at the moment, except when converting it to an `org.reactivestreams.Publisher` 😉

```scala
sealed trait Iterant[F[_], A] {

  // ...
  def toReactivePublisher(implicit F: Effect[F], ec: Scheduler): Publisher[A]
}
```

The only reason why `Effect` is needed here is because `Publisher` does not get evaluated in the `F[_]` context.

There are currently no other operations on `monix.tail.Iterant` requiring `Effect`. You can [checkout the implementation](https://github.com/monix/monix/blob/master/monix-tail/shared/src/main/scala/monix/tail/Iterant.scala) to see for yourself.

**Comment by alexandru** [comment]
And btw, I intend to keep `monix.tail.Iterant` free of `Effect` because if an `Effect` is required, that to me might be a design problem — note I am not speaking of FS2 which has other priorities.

**Comment by SystemFw** [comment]
doobie's `ConnectionIO` is also an `Async` but not an `Effect`, I think

**Comment by pchlupacek** [comment]
ok, I hear you. Also very short one. Why `shift` is on `Async`?

**Comment by alexandru** [comment]
> Also very short one. Why shift is on Async?

I don't know, personally I don't like it because it requires an `ExecutionContext` and while OK-ish for `IO.shift`, putting that in the type class is fishy.

I was thinking of adding a pure [data type](https://github.com/typelevel/cats-effect/issues/116) that describes such operations, but that's a long way coming and in order to safely expose it, we kind of need a cancelable `IO` or otherwise it wouldn't be safe — thinking about delays and shifts is what made me realize that it would be best for the standard `IO` to be cancelable.

**Comment by pchlupacek** [comment]
ok cool. We may then put shift to `Effect`, then `Async` will be solely capturing Asynchronous computations, and `shift` could be implemented with effect as it is on `Effect` and as such we may win on both sides?

**Comment by alexandru** [comment]
What do you mean?

Note `shift` as an operation isn't the problem per se, I used `shift` as an example as being the typical asynchronous computation that users want to wrap when using the `async` builder.

**All** async computations can be described as `(Either[Throwable, A] => Unit) => Unit`, this is the definition of asynchronous processing and this is what `Async` can wrap. Shove an `IO` or an `F[_]` in that callback and it is no longer quintessential.

**Comment by alexandru** [comment]
Sorry, closed by mistake 😀

**Comment by pchlupacek** [comment]
@alexandru no problem. I think the misunderstanding could be where to put `F`. I am just thinking that this will work: 
```scala
def async[A](k: (Either[Throwable, A] => Unit) => F[Unit]): F[A]

// then 

def shift(implicit ec: ExecutionContext): F[Unit] = {
    async { (cb: Either[Throwable, Unit] => Unit) =>
      F.delay{ 
        ec.execute(new Runnable {
          def run() = cb(Right(()))
        })
      }
    }
  }

```

Having said that I still believe shift has more home at `Effect` than on async :-)

**Comment by alexandru** [comment]
This is replaying to @SystemFw on Gitter, talking about the road to 1.0 — might as well respond here, since Gitter gets lost ...

> I'm becoming increasingly convinced that we need to add a version of `async` that takes an `(A => Unit) => F[Unit]` in addition to the one we have now that take `(A => Unit) => Unit`

I feel your pain there, but I've thought about it and knowing the implementation, I'm very convinced that we shouldn't.

From a usability point of view, you want `F[Unit]` in there to get rid of your `Effect` dependency in some cases, however that `Effect` dependency will be more uncommon going forward anyway.
The reason for why I don't want that is because increased flexibility at your call site comes at the expense of increased complexity in the implementation, as you're moving that responsibility around.

Basically if the implementation receives an `F[Unit]` from `Async`, now it has to execute it in its run-loop, instead of just ending the run-loop immediately in order to wait for the continuation. From the user's point of view, things are easy, you're passing an `F[Unit]` that may as well be synchronous, like a function, but internally you have to take into account the possibility that the received `F[Unit]` is also async.

```scala
def specialAsync[F[_], A](k: (A => Unit) => F[Unit])(implicit F: Effect[F]): F[A] =
  F.async { cb =>
    val io = F.runAsync(k(cb)) {
      case Left(e) => 
        // N.B. this call is concurrent and might violate the contract
        IO(cb(e)) 
      case Right(_) => 
        IO.unit
    }
    io.unsafeRunSync
  }
```

Such logic doesn't vanish, it only shifts to the implementation side and I'm very unconvinced that such an increase in complexity is worth it — I know for example that I don't want to do it in Monix's Task and that would be an *unfair tax* just for implementing `Async`.

Also, `Iterant` now implements `Async`, see https://github.com/monix/monix/pull/600 — I don't know if that ever will be useful, but by introducing `F[Unit]` in `async` we are effectively adding extra restrictions that were not there before. Because obviously, that `F[Unit]` no longer works as intended if `F` is a stream.

By bringing over the responsibility of `Effect`, we are bringing over its restrictions as well. `Effect` also cannot be implemented by streams.

Of course, we do not care much about streams implement `Async`, after all we have no current use-case other than being a cool curiosity, but we can't be sure that this doesn't introduce unintended consequences — we already had problems implementing `Effect` for `OptionT` and the `EitherT` instance is forced to be specialised for `Throwable`.

So we might not care for streams implementing `Async` right now, but we might discover that interesting types that we do want are no longer possible. Because when you're thinking `F[Unit]` in there, you're thinking `IO`, but `F` in that `Async` is not necessarily `IO`.

This is precisely why in the signature of `Concurrent#cancelable` I've used `IO[Unit]` for the cancelable token and not `F[Unit]`, realising that that is the correct approach after hitting myself against the wall of the `T` instances that we need to support and for which shoving an `F[Unit]` in there would be extremely complicated — as you're no longer talking about an underlying `IO[Unit]` but about an underlying `IO[Either[L, Unit]]` or about an underlying `IO[(S, Unit)]`, etc, and note usage of `IO` here is the happy path 😉

**Comment by alexandru** [comment]
Also guys, requiring `Effect` is usually a symptom — for example when you need to express `start` in your code, you need `Effect` indeed.

But the conclusion from that should not be to move that responsibility in `async` imo, but to express a generic `start`, because the underlying problem is that we need a `start` in the type class hierarchy. And expressing operations with `Effect` is what you can usually do when you don't have those operations in the super types of `Effect`.

**Comment by SystemFw** [comment]
So, the reason I'm mentioning this is that I'm afraid that even with `start` and `Concurrent`, we are going to need a constraint on `Effect`/`ConcurrentEffect` without a pure `async.
`start` will let us get rid of  using `runAsync` to implement `fork`, but there is another thing I'm not yet sure how to solve.

Basically in Promise we implement semantic blocking on an empty Promise. To do that, when you call `Promise.get`, we store a callback, and use `async` to have the `F` only return when the callback is called.

The problem is that the action we pass to `async` needs to modify a `Ref`, which is a pure operation in `F`, so to make it fit into `async` we need to `runAsync.unsafeRunSync` it.

Note: please consider though that I haven't actually tried hard to implement this with the new hierarchy, it might be that `start` alone is enough, this is just the impression I got (which is why I had mentioned it on Gitter only)

**Comment by alexandru** [comment]
@SystemFw @mpilquist @pchlupacek hi guys,

Just dropping a note that my opinion above is an educated one, but I'm not entirely sure.

Therefore I've decided to investigate how feasible the lazy `async` and `cancelable` builders are — in case they can be easily derived for the current monad transformers and in case `Async` still works for `monix.tail.Iterant`, then maybe we can squeeze them in.

But no promises and given that this breaks binary compatibility, it can't make it into 0.10.

Plus I'd advise you to accept `AtomicReference` into your life 😜 those are concurrency primitives, it's OK to be a little dirty.

**Comment by alexandru** [comment]
@SystemFw after your experience with migrating FS2 to 0.10, don't forget to add some feedback on whether this is still needed or not.

I haven't managed to play with it.
Scheduling a decision to be made on this one for 1.0.0-RC 🙂

**Comment by SystemFw** [comment]
@alexandru So, we kind of made without: Promise is implemented directly from scratch, and Promise in a way embodies the same concept as async with IO - semantic waiting on an async value.

I'd still like to have it, but only if there aren't any significant problems implementing it (which there might be as the impl would need to run the action, which is why implementing it in user code requires `ConcurrentEffect`).

@pchlupacek @mpilquist  thoughts?

**Comment by pchlupacek** [comment]
@SystemFw my position sort of stands. I always prefer _ => F[Unit] over _ => Unit :-) this always lead in much better reuse and code. So if we could implement it, then still +1

**Comment by alexandru** [comment]
Unfortunately I did not have time to play with an implementation to be able to assess its impact. If it's cheap (in the sense that it doesn't seem to add extra restrictions on the data types implementing `Async`, so all available instances remain valid), then it might make it.

**Comment by alexandru** [comment]
Hi all, so I played around with trying to define an `asyncF`. Unfortunately, as I feared, it extra restrictions over `Async`.

As an example, trying to implement it for `StateT[F, S, ?]` would go like this:

```scala
def asyncF[A](k: (Either[Throwable, A] => Unit) => StateT[F, S, Unit]): StateT[F, S, A] =
  StateT.liftF(F.asyncF(cb => k(cb).runEmptyA))
```

However doing that we also need a `cats.Monoid[S]` otherwise `runEmptyA` doesn't work.
And in fact we have the `cats.Monoid[S]` restriction on on the `Effect[StateT[F, S, ?]]` instance, because we have to do `fa.runA(S.empty)` there as well.

So that's a clear restriction that would be added on the `Async[StateT[F, S, ?]]` instance, a restriction that shows such an `asyncF` gets us much closer to `Effect`. Or in other words, by adding `asyncF`, the type class becomes less generic.

So we cannot add this operation on `Async`. If we ever feel the need for it, I think it should be on a different type class. It's still more generic than `Effect`, definitely, so maybe we'll do it at some point.

But for 1.x, I wouldn't do it unless we've got deal breakers in the projects depending on `cats-effect`. And given that you haven't cried out for it while updating FS2, I'm confident that we don't need it now 🙂
